src/function/maybe_changed_after.rs

```diff
@@ -268,15 +272,34 @@ where
             "{database_key_index:?}: validate_provisional(memo = {memo:#?})",
             memo = memo.tracing_debug()
         );
-        if (&memo.revisions.cycle_heads).into_iter().any(|cycle_head| {
-            zalsa
+        for cycle_head in &memo.revisions.cycle_heads {
+            let kind = zalsa
                 .lookup_ingredient(cycle_head.database_key_index.ingredient_index())
-                .is_provisional_cycle_head(
+                .cycle_head_kind(
                     db.as_dyn_database(),
                     cycle_head.database_key_index.key_index(),
-                )
-        }) {
-            return false;
+                );
+            match kind {
+                CycleHeadKind::Provisional => return false,
+                CycleHeadKind::NotProvisional => {
+                    // FIXME: We can ignore this, I just don't have a use-case for this.
+                    if C::CYCLE_STRATEGY == CycleRecoveryStrategy::FallbackImmediate {
+                        panic!("cannot mix fixpoint and `FallbackImmediate` cycles")
+                    }
+                }
+                CycleHeadKind::FallbackImmediate => match C::CYCLE_STRATEGY {
+                    CycleRecoveryStrategy::Panic => {
+                        // Queries without fallback are not considered when inside a cycle.
+                        return false;
+                    }
+                    // FIXME: We can do the same as with `CycleRecoveryStrategy::Panic` here, I just don't have
+                    // a use-case for this.
+                    CycleRecoveryStrategy::Fixpoint => {
+                        panic!("cannot mix fixpoint and `FallbackImmediate` cycles")
```

Comment:
I think this message could be clearer for an end user if it were framed in terms of the public API (that is, `cycle_result` vs `cycle_fn`).